<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta content="text/html;charset=ISO-8859-1"
 http-equiv="Content-Type">
<!-- Copyright 2007, 2008 Nanorex, Inc. See LICENSE file for details. -->
  <title>Nanorex NanoEngineer-1 Version 1.1.1</title>
</head>
<body>
<h1>Nanorex NanoEngineer-1 Version 1.1.1</h1>
NanoEngineer-1 (NE1) is an open-source (GPL) 3D multi-scale modeling
and simulation program for nano-composites with special support for
structural DNA nanotechnology. It features an easy-to-use interactive
3D graphical user interface for designing and modeling large,
atomically precise composite nanosystems.
<p>This is NanoEngineer-1 Version 1.1.1 as of July 16, 2008
</p>
<p>Click the following links to learn more about NanoEngineer-1
Version 1.1.1:
</p>
<ul>
  <li><a href="#SystemRequirements">System
Requirements</a> </li>
  <li><a href="#OfficialPlatforms">Officially Tested
Platforms</a> </li>
  <li><a href="#InstallationInstructions">Installation
Instructions</a> </li>
  <li><a href="#GettingStarted">Getting Started with
NE1</a> </li>
  <li><a href="#Extensions">Extensions for NE1</a>
  </li>
  <li><a href="#NewFeatures">New Features</a> </li>
  <li><a href="#KnownBugs">Known Bugs</a> </li>
  <li><a href="#Copyright">Copyright and Licensing</a>
  </li>
  <li><a href="#NewFeaturesArchives">New Features
Archives</a> </li>
</ul>
<a name="SystemRequirements"></a>
<h2>System Requirements</h2>
<h3>Minimum</h3>
<ul>
  <li>2.5 GHz Intel Pentium 4 CPU or equivalent </li>
  <li>Windows XP, Mac OS X 10.3.9, Linux </li>
  <li>1 Gb RAM </li>
  <li>100 Mb available disk space </li>
  <li>OpenGL compatible 3D video card </li>
</ul>
<h3>Recommended</h3>
<ul>
  <li>3 GHz Intel Core 2 CPU or equivalent </li>
  <li>Windows XP, Mac OS X 10.3.9, Linux </li>
  <li>2 Gb RAM </li>
  <li>100 Mb available disk space </li>
  <li>OpenGL compatible 3D video card with FBO and VBO support (<a
 href="http://www.delphi3d.net/hardware/extsupport.php?extension=GL_EXT_framebuffer_object">this
web page</a> gives a good list of recommended video cards.) </li>
</ul>
<a name="OfficialPlatforms"></a>
<h2>Tested Platforms</h2>
NanoEngineer-1 Version 1.1.1 has been tested on:
<ul>
  <li>Windows XP </li>
  <li>Mac OS X 10.4.9 (Tiger), and 10.5.2
(Leopard) </li>
  <li>Kubuntu 7.10 (Gutsy) </li>
</ul>
Limited testing has been done on:
<ul>
  <li>Windows Vista - it seems to work fine on Vista Business,
and with some successes on Vista Home </li>
</ul>
<a name="InstallationInstructions"></a>
<h2>Installation Instructions</h2>
<h3>Windows</h3>
Run the NanoEngineer-1 self-extracting installer
"NanoEngineer-1_1.1.1.exe".
<h3>Mac OS X</h3>
Run the NanoEngineer-1 self-extracting installer
"NanoEngineer-1_1.1.1.dmg".
<p>Note: If during installation, the message: "You cannot
continue. There is nothing to install." appears, navigate to
&lt;your hard drive&gt;/Library/Receipts/ and delete any files
that start with "NanoEngineer-1". Then empty the Trash and run the
installer again.
</p>
<h3>Linux and Developers</h3>
NanoEngineer-1 Version 1.1.1 must be compiled from sources for Linux
users and developers. Download the NanoEngineer-1 Version 1.1.1 source
distribution "NanoEngineer-1_1.1.1.tar.gz" and follow the instructions
found in the following steps:
<ol>
  <li> <a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Build_Requirements">NE1
Build Requirements</a> - Before building NE1, check that the
requirements for your platform are met. This page lists all the
libraries necessary to build NE1, including some notes on library
installation. </li>
  <li> <a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Build_and_Install">NE1
Build and Install</a> </li>
</ol>
<a name="GettingStarted"></a>
<h2>Getting Started with NE1</h2>
The best way to get started using NanoEngineer-1 is:
<ol>
  <li>Peruse this <a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=Introduction_to_the_NanoEngineer-1_User_Interface">Introduction
to the NanoEngineer-1 User Interface</a>. </li>
  <li>Then work through some of the tutorials on our <a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=Tutorials">Tutorials</a>
wiki page. </li>
</ol>
<a name="Extensions"></a>
<h2>Extensions for NE1</h2>
NE1 makes use of a number of free extensions and companion
applications, including:
<ul>
  <li>GROMACS an extremely high performance, open source,
and cross-platform package used to perform molecular dynamics
simulation of systems with hundreds to millions of particles.&nbsp;
  </li>
  <li>Open Babel, used to import/export different chemistry file
formats to/from NE1. </li>
  <li>QuteMolX, a high-quality molecular visualization system
that can be "launched and loaded" from NE1. Very cool!<!-- <li>NanoVision-1, an application for viewing molecular structures and simulations, especially those produced by NanoEngineer-1. -->
  </li>
  <li>POV-Ray, which creates high-quality renderings of the
current model with a single click of the mouse. </li>
  <li>MegaPOV (for Mac OS X), which creates high-quality
renderings of the current model with a single click of the mouse.
POV-Ray is also required. </li>
  <li>GNUPlot, used to plot the results of molecular dynamics
simulations using NanoDynamics-1, NE1's own molecular dynamics engine
for nanomechanical simulations. It is installed with NE1 for Windows
and Mac. </li>
</ul>
For more details on all extensions and links to download pages, visit: <a
 href="http://nanoengineer-1.com/dev1/index.php?option=com_content&amp;task=view&amp;id=31&amp;Itemid=86">
NanoEngineer-1 Extensions</a>
<p>Note that the <b>NanoEngineer-1 Application Suite</b>
includes all of the above extensions and companion applications in one
convenient installer. You can find it at <a
 href="http://www.nanorex.com/">www.nanorex.com</a>.
<a name="NewFeatures"></a></p>
<h2>New Features</h2>
<b>What is new in NanoEngineer-1 Version 1.1.1 (7/16/2008):</b>
<ol>
  <li> User Interface Modifications&nbsp;&nbsp;
    <ul>
      <li>Added a preferences option to display or hide the
Confirmation Corner.</li>
      <li>"Escape" key exits current command.</li>
    </ul>
    <ul>
      <li>Added Color combobox option to DnaSegment and DnaStrand
PMs.</li>
      <li>Added DNA Cylinder to View &gt; Display menu.</li>
      <li>Added DNA Display Style to Edit &gt; DNA Display
Style menu.</li>
      <li>Improved color contrast between the background color
and line
color of region selection polyline (rectangle/lasso), DNA lines and
Nanotube lines.</li>
      <li>New disc geometry for DnaSegment resize
handles&nbsp;to aid resize operations.</li>
      <li>Enlarged the resize handles (the 3D arrowhead
only)&nbsp;when resizing a Dna Segment</li>
    </ul>
  </li>
  <li>Other new features</li>
  <ul>
    <li>Plane images are saved/restored between sessions.</li>
    <li>NanoEngineer-1 is now built with Qt 4.3.5.</li>
    <li>IOS import and export feature added</li>
    <li>Added the ability to draw a dna duplex on specified plane</li>
  </ul>
  <li> Fixed Bugs
    <ul>
      <li>Problem with a properties manager font size
inconsistency<br>
      </li>
      <li>OpenBabel export now executes correctly<br>
      </li>
      <li>OpenBabel export no longer zooms the screen after
export.<br>
      </li>
      <li>Miscellaneous screen and menu layout problems
concerning the switch to Qt 4.3.5</li>
      <li>Modified the file selection box extension choices for
OpenBabel export/import so that file extensions are displayed on Mac</li>
      <li>now its possible to select a plane by clicking anywhere
inside it</li>
      <li>Fixed bug where the bases per turn value was incorrect</li>
      <li>Dna Segment property manager could display the wrong
number of base pairs</li>
      <li>When resizing a Dna Segment 1, the cursor text that
indicates length delta is off by 1</li>
      <li>Resizing a Dna Segment adds one base pair when finished</li>
      <li>Fixed a bug where the "Bases per turn", duplex rise
widgets in the Insert DNA don't honor global preferences.</li>
      <li>Other miscellaneous bug fixes</li>
    </ul>
  </li>
</ol>
<br>
<br>
<hr><br>
See <a href="#NewFeaturesArchives">New Features Archives</a>
for information on previous versions.
<a name="KnownBugs"></a>
<h2>Known Issues and Bugs</h2>
Currently, NanoEngineer-1 and its installers should work properly on
MacOSX 10.4+.
Due to a switch to a newer version of the
installer system, the NanoEngineer-1 folder may be installed directly
into the toplevel folder of the root partition, rather than into its
intended location of /Applications/Nanorex. &nbsp;During testing,
this error did not occur on versions of MacOSX 10.4 and above.
&nbsp;If
this occurs, just move the entire NanoEngineer-1 folder to its intended
location
using the Finder.
The application itself may not work on MacOSX 10.3 as it did before,
due to a crash apparently related to our use of a newer version of Qt
(4.3.5).
You may access a current copy of NE1 Known Bugs on-line at:
<p><a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Known_Bugs">http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Known_Bugs</a>
<a name="Copyright"></a></p>
<h2>Copyright and Licensing</h2>
NanoEngineer-1 Copyright 2004-2008 Nanorex, Inc.
<p>This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.
</p>
<p>This program is distributed in the hope that it will be
useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.
</p>
<p>You should have received a copy of the GNU General Public
License along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
USA.
<a name="NewFeaturesArchives"></a></p>
<h2>New Features Archives</h2>
<b>What is new in NanoEngineer-1 Version 1.1.0 (6/4/2008):</b>
<ol>
  <li> User Interface Modifications&nbsp;&nbsp;
    <ul>
      <li>Resize multiple DnaSegments at once.</li>
    </ul>
    <ul>
      <li>DNA cursor text display options in a new PM group box.
Display options (as checkboxes): Bases, Turns, Length, Angle.</li>
      <li>DNA and Nanotube cursor text should auto-switch b/w
Angstroms (0-10 A) and nanometers ( &gt; 10 A).</li>
      <li>atom overlap indicator - quick access to a toggle for
it in the UI</li>
      <li>Edit DNA Display Style property manager that allows the
user to edit and preview DNA display options.</li>
      <li>Order DNA property manager modifications.</li>
      <li>Make Crossovers command (initial implementation, for
PAM3 model)</li>
      <li>Colored wire-frames as an option for hover highlighting
and selection.</li>
      <li>Cleaned up the Run Dynamics and Minimize Energy dialogs.</li>
    </ul>
  </li>
  <li>Other new features</li>
  <ul>
    <li>NE1 installers associate files with the .mmp extension
with NE1 for all platforms.</li>
    <li>GROMACS 3.3.3 build (double precision)</li>
    <li>Better support for languages which have characters which
do not show up in the ASCII character set</li>
    <li>Added the ability to display an image on a reference plane</li>
    <li>Added support for stereo image displays.</li>
    <li>Added support for fog, accessible from Color Scheme PM
and the Preferences dialog.</li>
  </ul>
  <li> Fixed Bugs
    <ul>
      <li>Hover highlighting not working after mouse wheel
zooming.</li>
      <li>reduce minimum size of User Prefs Dialog widget</li>
      <li>fixed bug in breaking a PAM5 crossover bond</li>
      <li>other minimizer issues listed at PAM model tasks for
v1.1.0</li>
      <li>fixed ND1 PAM3 minimize for crossovers</li>
      <li>fixed bug&nbsp;in strand list widget selection or
update, exposed or caused by ghost bases being in same DnaGroup as
their axis</li>
      <li>spontaneous PAM5 conversion on duplex resize</li>
      <li>misc. necessary features to speed development and
testing of PAM force field</li>
      <li>fix major bug in saving a PAM5 ring duplex</li>
      <li>revise Minimize Energy in the following ways:</li>
      <ul>
        <li>always "Minimize All" by default; nonempty selection
enables (but does not check) checkbox for "Minimize Selection"</li>
        <li>Minimize Selection passes *all* atoms (not just those
near enough to the selection), anchoring all unselected atoms</li>
      </ul>
      <li>leave out PAM bondpoints instead of writing H in place
of them</li>
      <li>PAM sugar atoms aren't connected when a dna segment is
resized by 1 basepair</li>
      <li>Cancel button to abort reading a huge file doesn't work</li>
    </ul>
  </li>
</ol>
<h2><b>PAM5 and PAM3 DNA Force Field Improvements (for
v1.1.0)</b>
</h2>
Both the PAM5 and PAM3 (pseudo-atom reduced model, 5 and 3 pseudo-atom
structure) DNA force fields have undergone refinement.
<br>
<br>
<b>1. PAM5-Specific Improvements</b>
<br>
<br>
This version of NE1 contains an alpha release of the PAM5 (pseudo-atom
model, 5 pseudo-atom structure) force field for the modeling of
DNA-based nanostructures.<span class="Apple-converted-space">&nbsp;
</span>This force field is being provided prior to publication
for testing and benchmarking purposes.<span
 class="Apple-converted-space">&nbsp; </span>While
further refinement is likely to follow, the current version has been
found to be useful in the modeling of several classes of DNA
nanostructures and is not expected to undergo significant modification.
<br>
<br>
The force constants generated from the parameterization process are the
result of comparisons with mutually inconsistent sources.<span
 class="Apple-converted-space">&nbsp; </span>The
bending stiffness values suffer most from the available selection of
data sets used for the force constant development.<span
 class="Apple-converted-space">&nbsp; </span>Consequently,
the relative energies of strained structures should not (yet) be
considered meaningful.<span class="Apple-converted-space">&nbsp;
</span>The beta version will fit stretch, bend, and twist modulii
(as well as the stretch-twist cross term and several others) based on a
combination of experimental and molecular dynamics (published and
currently &ldquo;in-house&rdquo;) results.
<br>
<br>
More information about the PAM5 force field, including a description of
the interhelical potential and example structures, is available at:<span
 class="Apple-converted-space">&nbsp;</span>
<br>
<br>
<a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=PAM3_and_PAM5_Model_Descriptions"><span
 class="s1">http://www.nanoengineer-1.net/mediawiki/index.php?title=PAM3_and_PAM5_Model_Descriptions</span>
</a><br>
<br>
<b>2. Interhelical Description</b>
<br>
<br>
The screened Coulombic (Yukawa) potential used to model the
interhelical interactions in PAM5 has been implemented as an additional
energy minimization option for the PAM3 force field.<span
 class="Apple-converted-space">&nbsp; </span>The
Yukawa potential is used in place of the Coulombic potential on the
backbone pseudo-atoms.<span class="Apple-converted-space">&nbsp;
</span>In PAM3, the charged pseudo-atoms mark the sugar positions.<span
 class="Apple-converted-space">&nbsp; </span>In
PAM5, the charged pseudo-atoms mark the phosphate positions.
<br>
<br>
The constants in the Yukawa potential have the following default
settings (which can be adjusted to new default settings in the
sim-params.txt file):
<br>
<ul>
  <li>Counter-ion Charge: 2.0 (Mg<sup>2+</sup>
counter-ion) </li>
  <li>Counter-ion Molarity: 0.02 M </li>
  <li>System Temperature: 298.0 K </li>
  <li>Dielectric Constant: 78.5 (water) </li>
</ul>
<br>
The Yukawa potential has the following general influences on structure
optimizations:
<br>
<ul>
  <li>Increased separation (&ldquo;bowing&rdquo;) between
neighboring duplexes (duplex-duplex repulsion) </li>
  <li>Increased separation between free duplex ends (those not
connected by
crossovers) at tight convergence (EndRMS &lt; 5.0 pN) </li>
  <li>Curvature at the corners of three-dimensional DNA
structures (a
result of local repulsive interactions at junction points) </li>
</ul>
The screened Coulombic (Yukawa) potential is always repulsive as
currently implemented.<span class="Apple-converted-space">&nbsp;
</span>In atomistic simulations containing discrete counter-ions,
duplexes can be bridged by these counter-ions to provide local binding
stability.<span class="Apple-converted-space">&nbsp; </span>This
leads to duplex loops with smaller average separations
(&ldquo;bowing&rdquo;) and duplex ends in crossover junctions
that do not separate.<span class="Apple-converted-space">&nbsp;
</span>As a screened Columbic potential applied to the phosphate
(PAM5) or backbone sugar (PAM3) pseudo-atoms, the Yukawa potential
provides no such local binding between duplexes.
<br>
<br>
<b>3. PAM3/PAM5 Energy Minimization Issues and Procedures</b>
<br>
<br>
A general procedure for modeling DNA nanostructures using the PAM5
force field is as follows:
<br>
<br>
<ol>
  <li>Draw the components using the PAM3 structures in the
&ldquo;Build DNA&rdquo; Mode </li>
  <li>Perform an Energy Minimization using PAM3 </li>
  <ul>
    <li>With or
Without Electrostatics (the Yukawa potential) </li>
    <li>With or
Without Neighbor Searching (regeneration of the atom pair table for the
calculation of electrostatic interactions.<span
 class="Apple-converted-space">&nbsp; </span>This
slows down the energy minimization significantly but leads to more
accurate structures) </li>
  </ul>
  <li>Confirm the Optimized Structure minimized correctly</li>
  <ul>
    <li>If not,
adjust the structure and perform an additional Energy Minimization </li>
  </ul>
  <li>Convert the PAM3 structure to PAM5 with the &ldquo;PAM3
to
PAM5&rdquo; Button </li>
  <li>Perform an Energy Minimization using PAM5 </li>
  <ul>
    <li>With or
Without Electrostatics (the Yukawa potential) </li>
    <li>With or
Without Neighbor Searching (regeneration of the atom pair table for the
calculation of electrostatic interactions.<span
 class="Apple-converted-space">&nbsp; </span>This
slows down the energy minimization significantly but leads to more
accurate structures) </li>
  </ul>
  <li>Confirm the Optimized Structure minimized correctly </li>
  <ul>
    <li>If not,
adjust the structure and perform an additional Energy Minimization </li>
  </ul>
</ol>
<br>
The PAM3 force field contains fewer pseudo-atoms than PAM5 and
relatively stiff bending terms.<span class="Apple-converted-space">&nbsp;
</span>This makes energy minimization faster with PAM3,
especially when the starting structure is significantly different than
the optimized (equilibrium) structure.<span
 class="Apple-converted-space">&nbsp; </span>As a
general rule, the minimization procedure for a new structure should
begin with PAM3 prior to the more accurate, but more time consuming,
PAM5 minimization.
<br>
<br>
<b>What is new in NanoEngineer-1 Version 1.0.1 (5/6/2008):</b>
<ol>
  <li> User Interface Modifications
    <ul>
      <li> New selection/highlighting/drag behavior in Build Dna
and its sub-commands.. </li>
      <li> Edit DnaStrands and DnaSegments on single-click. </li>
      <li> Option to render 3' and 5' arrowheads/spheres in
different color. </li>
      <li> Convert bonds in "Build Chunks" command. </li>
      <li> New Dna seletion/highlighting and drag behavior in all
modes. See Dna Selection and Movement perations. </li>
    </ul>
  </li>
  <li> Fixed Bugs
    <ul>
      <li>DNA Cylinder display style: Axis should not be
displayed on ssDNA segments. </li>
      <li>Double-clicking strand does not select mate strands in
MT. Hiding all strands only shows double-clicked strand as hidden in
MT. </li>
      <li>undo/redo doesn't restore selectedness of DnaStrands </li>
      <li>Joining strands doesn't always pull in all necessary
chunks (except for the dna updater). </li>
      <li>Closing or "tearing-off" history makes the energy
minimize fail. </li>
      <li>Tooltip shows the Windows hotkey even on Mac. </li>
      <li>Fix Default path for all NE1 plug-ins. </li>
      <li>Zoom to proper scale (i.e. 50A half height) when
editing a duplex the first time. </li>
      <li>Button images are missing from the sequence editor on
Mac, an installer issue. </li>
      <li>Strand resize handle color is not updated for new
strand. </li>
      <li>Atoms not rendered when global display style is "Dna
Cylinder". </li>
      <li>Default settings for DNA Cylinder display style. </li>
      <li>Length of sequence unknown when typing into sequence
editor. </li>
      <li>Tooltips and Help docs should include keyboard
shortcuts. </li>
      <li>Bug when a strand is lengthened by 1 strand base atom </li>
      <li>Hitting enter in sequence editor inserts newline (which
appears to clear the sequence) and inserts an "N" into the mates field.
      </li>
      <li>Mao 3-point star has discrepancies. </li>
      <li>SDN structure clean up in part library. </li>
    </ul>
  </li>
</ol>
<b>What is new in NanoEngineer-1 1.0.0 Version (4/25/2008):</b>
<ol>
  <li> Core User Interface Operations Support
    <ul>
      <li> A new DNA data model has been implemented that makes
use of a unique DNA structure checking mechanism. </li>
    </ul>
  </li>
  <li> User Interface Improvements and Additions
    <ul>
      <li> DNA Strand Operations
        <ul>
          <li> New handles for resizing segments and strands
along the axis. </li>
          <li> Nick/break/join strands - now you can break a
strand or join two strands via a break/join command on the toolbar. </li>
          <li> Create a crossover in one simple operation by
selecting nucleobases from adjacent strands and invoking the context
menu. </li>
        </ul>
      </li>
      <li> DNA Segment Operations
        <ul>
          <li> Change the length of segment
            <ul>
              <li> Constrained to central axis </li>
              <li> By adding base pairs </li>
            </ul>
          </li>
          <li> Rotate a segment about its central axis </li>
          <li> Translate a segment along its central axis </li>
        </ul>
      </li>
      <li> DNA Duplex Operations
        <ul>
          <li> Create a single DNA duplex by selecting two points
on a plane
            <ul>
              <li> The duplex is shown while you're selecting the
second point and can be displayed in either ladder or ribbon display
styles. </li>
              <li> The duplex adheres to intelligent horizontal
and vertical "snap-to-grid" rules allowing for rapid, horizontal and
vertical placement. </li>
              <li> Text next to the mouse cursor dynamically
updates the number of base-pairs and the length of the duplex as you
change the rubberband line during DNA duplex creation. </li>
            </ul>
          </li>
        </ul>
      </li>
      <li> Sequence Editor
        <ul>
          <li> Load sequences from a text file, paste them in, or
even type them. </li>
          <li> Unique find/replace feature lets you quickly find
and replace sub-sequences. </li>
          <li> Easily switch between 5'-3' and 3'-5' orientation.
          </li>
          <li> Save sequences to a separate text file. </li>
        </ul>
      </li>
      <li> Peptide Generator
        <ul>
          <li> This experimental module can create atomistic
models of polypeptide chains of arbitrary sequence and secondary
structure. </li>
          <li> The polypeptide chain can be built step-by-step by
choosing consecutive amino acids from a set of the standard 20
naturally occurring amino acids. </li>
          <li> Local geometry of the built chain can be chosen
from a set of predefined secondary structure geometries (alpha helix,
3-10-helix, pi-helix, or beta-strand) or explicitly defined by
specifying a pair of phi/psi angles. </li>
          <li> A sequence of the peptide is visualized in a
sequence viewer window and it is colored by current secondary structure
type. </li>
          <li> Note:
            <ul>
              <li> Generated structures can't be minimized or
simulated in a reliable way because the ND1 force field currently lacks
certain atom types (e.g. amide-type nitrogen). The generated geometry
is idealized, but after minimization it becomes wrong. </li>
              <li> The structure is not visualized as it is being
built, you need to press the "Preview" button to see it. </li>
              <li> The structure can't be modified after it has
been generated, nor can the sequence be edited. </li>
            </ul>
          </li>
        </ul>
      </li>
      <li> Miscellaneous Improvements and Additions
        <ul>
          <li> DNA generator now supports PAM3 and PAM5 models,
but <b>not</b> atomistic models. </li>
          <li> A new rainbow-effect cylinder display style is now
available for viewing reduced model PAM DNA structures. </li>
          <li> A new nanotube generator - create a single-walled
carbon nanotube by selecting two points on a plane. </li>
          <li> Toolbar and dockwidget positions can now be saved
between sessions. </li>
          <li> Option to maintain the aspect ratio while resizing
a plane. The ratio is updated automatically (i.e. anytime the plane's
width and/or height are changed). </li>
          <li> The bond offset sphere checkbox in Extrude
property manager is back. </li>
          <li> Mirror feature enhancements - it can now mirror
everything but the planes. </li>
          <li> Triple-click any PAM atom to select all connected
PAM atoms. </li>
          <li> New paste mode and related enhancements
            <ul>
              <li> New paste mode which can be accessed using
menu Edit &gt; Paste from clipboard... or corresponding button in
the standard toolbar. You can return to the previous mode by hitting
hitting the 'Done' or 'Escape' key. </li>
              <li> Clipboard groups are now 'pastable' (so those
are also included in the clipboard groupbox of Paste Mode) </li>
              <li> Edit &gt; Paste or Ctrl + V does a
'single-shot' paste operation. It pastes the most recently copied item.
The item placement position is slightly modified compared to the
original copied item. </li>
              <li> Improvements to group pasting via paste mode.
Now the groups are pasted at a closer place to where you
double-clicked. </li>
            </ul>
          </li>
          <li> New part library mode - the part library can be
accessed in any mode by using Insert &gt; Part from partlib...
menu. You can return to the previous mode by hitting the 'Done' or
'escape' key. </li>
          <li> When a single atom is selected, a new groupbox in
the Build Atoms Property Manager, Selection Options, becomes active and
shows the x, y, z coords of the selected atom. You can then use the
spinboxes to move the atom precisely. </li>
          <li> NE1 can now be shown in full-screen mode (only
menus and rulers appear around the work area) and semi-full-screen mode
(includes all the toolbars as well.) </li>
        </ul>
      </li>
    </ul>
  </li>
  <li> Structural DNA Nanotech (SDN) Part Library Additions
    <ul>
      <li> Several DNA double-crossovers such as DAE, DAO, DPE,
DPON, DPOW, DX Tile, DX+J Tile, JX2, and PX65 </li>
      <li> Several SDN structures such as Mao's 3- and 6-point
stars, Rothemund's star origami, Pierce's walker, Turberfield's
tetrahedron, and Shih's octahedron. </li>
    </ul>
  </li>
  <li> PAM5 and PAM3 Reduced DNA Models
    <ul>
      <li> The original PAM5 has undergone further refinements to
improve the realism of minimized SDN structures </li>
      <li> A new PAM3 model has been added for simplified
creation, and clear, attractive display of SDN motifs </li>
    </ul>
  </li>
  <li> QuteMol<sub>X</sub> Integration
    <ul>
      <li> QuteMol<sub>X</sub> is our branch of
QuteMol. The branch allows us to continue to improve the code, and
better integrate it with NE1. </li>
      <li> QuteMol<sub>X</sub> now matches the
display mode, background color, and atom coloring of NE1 when either
launched from NE1, or when opening PDB files exported from NE1. </li>
    </ul>
  </li>
  <li> GROMACS Integration
    <ul>
      <li> GROMACS has been added as an energy minimization
engine. Structures can be minimized in the foreground, or in a separate
background GROMACS process. The location of all the GROMACS
input/output files, and the identifier (PID) for the GROMACS process is
printed in the History text area at the bottom of NE1. An application
called NanoVision-1, which is currently in development and not yet
available, will be able to attach to, monitor, and cleanly abort the
spawned GROMACS processes. </li>
      <li> HDF5 support was added to GMX for writing and viewing
trajectory files, but won't be fully realized until NanoVision-1 is
available. </li>
    </ul>
<!--
<li> NanoVision-1
<ul>
<li> NanoVision-1 (NV1) is a cross-platform, open-source application that displays structures and simulations especially those produced by NE1 with GROMACS.
<li> NV1 launches automatically when a backgrounded GROMACS minimization is initiated and shows a visualization of the minimization as it progresses.
</ul>
--></li>
  <li> Notable Bug Fixes
    <ul>
      <li> The NE1 process now reliably ends upon exit of NE1. </li>
      <li> POV-Ray scene fixed - bonds are now the correct color
when a chunk is displayed in ball-and-stick style, and colored. </li>
      <li> PAM reduced DNA models are now rendered properly in
QuteMolX. </li>
      <li> Fixed the "Rotate as a unit" checkbox in the Rotate
groupbox. </li>
      <li> Moiety copies are now placed closer together when
extruding to a ring as final product. </li>
      <li> Inserting a part with a hotspot from the parts library
now works. </li>
      <li> Newly read mmp files are no longer immediately marked
as changed. </li>
      <li> Added a mouse wheel governor for Mac users - using the
mouse wheel to zoom is now manageable. </li>
    </ul>
  </li>
</ol>
<b>What is new in NanoEngineer-1 Version 0.9.1 (7/01/2007):</b>
<ol>
  <li> UI Improvements
    <ul>
      <li> New <a
 href="http://www.nanoengineer-1.net/mediawiki/index.php?title=Confirmation_Corner">Confirmation
Corner</a> </li>
      <li> Command Manager includes custom flyout toolbars for
all commands </li>
      <li> Property Manager includes helpful messages for <b>Plane</b>,
        <b>Build Atoms</b>, <b>Extrude</b> and
        <b>Translate/Rotate</b> commands </li>
      <li> <b>Build Atoms</b> includes the new <b>Cut
Bonds</b> bond tool for deleting bonds between two atoms </li>
      <li> Splitter between the left channel (i.e. the Model
Tree/Property Manager) and the 3D graphics window can be resized </li>
    </ul>
  </li>
  <li> DNA Generator
    <ul>
      <li> Specify DNA duplex by length (in Angstroms) </li>
      <li> Fixed some bugs involving the Strand Sequence text
edit widget (in the Property Manager) </li>
    </ul>
  </li>
  <li> <a href="http://qutemol.sourceforge.net/">QuteMol</a>
    <ul>
      <li> NE1 writes and QuteMol (0.4.1) reads and renders
bondpoints </li>
      <li> NE1 now writes the PDB REMARK records needed for
QuteMol (a future version) to render:
        <ul>
          <li> Initial view orientation </li>
          <li> Background color </li>
          <li> Launch display style (launch = "synchronize with
NE1 at startup") </li>
          <li> Launch tubes/licorice cylinder "thickness" </li>
          <li> Launch ball and stick cylinder "thickness" </li>
          <li> Chunks' display style and color </li>
        </ul>
Be warned, however, that QuteMol demands high-end graphics hardware.
Also:
        <ul>
          <li> QuteMol runs well on the "Pro" series of Mac
products, (i.e. Macbook Pro and Mac Pro), but it can hang other Macs
(i.e. MacMini, MacBook, and some other PowerPC based Macs). QuteMol has
not been known to hang Windows-based computers. </li>
          <li> Check <a
 href="http://www.delphi3d.net/hardware/extsupport.php?extension=GL_EXT_framebuffer_object">this
list</a> of supported graphics cards and drivers. </li>
          <li> If your computer includes a graphics card on this
list but still isn't rendering properly, then be sure to update your
graphics card driver. </li>
        </ul>
      </li>
    </ul>
  </li>
  <li> Reference Plane (<b>Insert &gt; Reference
Geometry &gt; Plane</b>)
    <ul>
      <li> New option <b>Offset to a Plane</b>
creates a new plane from an offset to an existing plane </li>
      <li> Edges are aligned with the screen when <b>Parallel
to Screen</b> option is selected </li>
      <li> Works with <b>Mirror</b> command </li>
      <li> Works for all <b>Translate</b> commands
(i.e. <b>To XYZ Position</b>) </li>
      <li> Preview improvements </li>
      <li> Many resizing bugs fixed </li>
    </ul>
  </li>
  <li> Other notable bug fixes:
    <ul>
      <li> The current working directory is set to the last
directory in which a file was opened or saved. This is persistent
between seesions </li>
      <li> The graphitic Nitrogen atomic hybrid can be deposited
and transmutes correctly </li>
      <li> <b>Simulation &gt; Play Movie</b> is
enabled only when there is a valid movie </li>
      <li> Fixed some button text overlapping and layout problems
in <b>Build Atoms</b> Property Manager </li>
    </ul>
  </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What is new in NanoEngineer-1 Version 0.9 (6/1/2007):</b>
</p>
<ol>
  <li> Completely ported GUI from Qt 3.x to Qt 4.x
    <ul>
      <li> NanoEngineer-1 Alpha 9 is now GPL'd for all platforms,
including Windows </li>
    </ul>
  </li>
  <li> Improved Graphical User Interface, improved ease of use
    <ul>
      <li> New Property Manager presents users with options for
current commands </li>
      <li> New Command Manager Toolbar provides easy access to
most common commands </li>
      <li> Improved selection and highlighting heuristics </li>
      <li> Better, more intuitive interactive graphics for
translating and rotating selected components </li>
      <li> Interactive zooming on cursor focus point </li>
    </ul>
  </li>
  <li> Special support for Structural DNA Nanotechnology
    <ul>
      <li> DNA "reduced model" allows easy design of complex DNA
structures </li>
      <li> Collection of DNA motifs included in part library:
        <ul>
          <li> Basic motifs (i.e. holliday junctions, double
crossovers, triple crossovers, etc.) </li>
          <li> Star structures </li>
          <li> DNA tiles </li>
          <li> DNA origami </li>
        </ul>
      </li>
    </ul>
  </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What is new in NanoEngineer-1 Version 0.8 (7/13/2006):</b>
</p>
<ol>
  <li> New Structure Generators
    <ul>
      <li> DNA Generator - creates custom DNA structures
        <ul>
          <li> B-DNA and Z-DNA </li>
          <li> Single or double stranded </li>
        </ul>
      </li>
      <li> Enhanced Nanotube Generator
        <ul>
          <li> Carbon or Boron Nitride nanotubes </li>
          <li> Multi-walled nanotubes </li>
          <li> Twisted option </li>
          <li> Bend option </li>
          <li> Z and XY distortions options </li>
        </ul>
      </li>
      <li> Heterojunction
        <ul>
          <li> Generates and connects two nanotubes of arbitrary
size. </li>
          <li> It uses using a C++ port of CoNTub, an algorithm
for connecting two arbitrary carbon nanotubes. </li>
        </ul>
      </li>
      <li> Graphene Generator
        <ul>
          <li> Creates a sheet of graphene given width and height
dimensions. </li>
        </ul>
      </li>
    </ul>
  </li>
  <li> POV-Ray Scene and Ray Trace scene
    <ul>
      <li> These provide a convenient way to preview the results
of a POV-Ray scene rendering in NanoEngineer-1. To use this feature,
you should install POV-Ray or MegaPOV and enable it as a plug-in. </li>
    </ul>
  </li>
  <li> Support for adding comments to the model </li>
  <li> New Chunk display
    <ul>
      <li> Chunks can be rendered in a new Cylinder display mode.
(Experimental feature) </li>
    </ul>
  </li>
  <li> Simulator dialog
    <ul>
      <li>Options to specify how often to update the screen
during the simulation. </li>
    </ul>
  </li>
  <li> Adjust All and Adjust Selection
    <ul>
      <li> Quickly adjusts positions of atoms and bonds to make
geometry of structures more realistic. </li>
    </ul>
  </li>
  <li> Minimize Energy
    <ul>
      <li> New dialog for providing fine grain control of
structure minimizations. </li>
    </ul>
  </li>
  <li> Simulator:
    <ul>
      <li> Torsion and out-of-plane terms in the simulator.
(Although these terms are included, please note that this still has
some known bugs and need more work. For example, minimize returning
high values for some structures or taking long time to finish) </li>
    </ul>
  </li>
  <li> Infer bonds, when reading PDB files that don't have any
bonds </li>
  <li> Build Mode improvements:
    <ul>
      <li> Bond types can be changed using context menu or bond
tools, even when this requires changing atom types or removing
bondpoints. </li>
      <li> The bond order of a bondpoint can be changed using
bond tools. </li>
      <li> Automatic repositioning of bondpoints when dragging
connected atoms has been improved </li>
    </ul>
  </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in NanoEngineer-1 Version 0.7 (4/24/2006):</b>
</p>
<ol>
  <li> Undo/Redo:
    <ul>
      <li> NanoEngineer-1 now includes rich Undo/Redo
functionality </li>
      <li> Automatic checkpointing can be enabled/disabled by
selecting 'Edit &gt; Automatic Checkpointing' </li>
      <li> 'Edit &gt; Clear Undo Stack' clears all Undo
checkpoints to free system memory </li>
    </ul>
  </li>
  <li> 'Select Atoms Mode' has been merged with 'Build Mode' </li>
  <li> Build Mode:
    <ul>
      <li> Significantly improved interactive mouse controls for
building, selecting, moving and editing structures and other objects.
To learn more about the new mouse controls, see the 'Build Mode'
section of the Mouse Control documentation by selecting 'Help &gt;
Mouse Controls...' </li>
      <li> Hover highlighting now supports bonds and jigs (not
just atoms) </li>
      <li> Atoms, bonds and jigs have context menus for common
commands </li>
      <li> Atom Selection Filter, which filters the selection of
one or more atom types when performing atom selection operations </li>
      <li> Transmute atoms </li>
      <li> New Library tab in the MMKit allows 3D preview and
insertion of parts from the NanoEngineer-1 library </li>
      <li> 'Water' selection filter can be enabled/disabled from
the dashboard (disabled by default) </li>
      <li> Build mode is the new default startup mode </li>
    </ul>
  </li>
  <li> On-Line Wiki Help:
    <ul>
      <li> Context sensitive help using the F1 key </li>
      <li> Users can contribute to and/or edit the On-Line Wiki
Help documentation </li>
    </ul>
  </li>
  <li> Molecular Dynamic Simulation:
    <ul>
      <li> Improved force field. Formal bend parameters have
replaced guessed values for all single bond combinations </li>
      <li> Warnings emitted when parameters are used that are not
determined by quantum calculations </li>
      <li> Warnings when bonds exceed reasonable limits during
simulation </li>
      <li> Watch minimize and MD simulations in real time </li>
      <li> Minimize and MD simulations can be stopped at any
time, leaving atoms at their current positions </li>
      <li> Improved minimization algorithm </li>
    </ul>
  </li>
  <li> View Menu/Toolbar:
    <ul>
      <li> 'Set View Normal To' sets the view to a new view based
on the normal vector of the current selection </li>
      <li> 'Set View Parallel To' sets the view to a new view
based on the parallel vector of the current selection </li>
      <li> 'Rotate View 180' rotates the current view 180 degrees
around the vertical axis </li>
      <li> 'Rotate View +90' rotates the current view 90 degrees
around the vertical axis </li>
      <li> 'Rotate View -90' rotates the current view -90 degrees
around the vertical axis </li>
      <li> 'Save Named View' saves the current view as a custom
view in the Model Tree. Using the context menu of the Named View icon
in the Model Tree, the view can be restored </li>
    </ul>
  </li>
  <li> Single Walled Carbon Nanotubes:
    <ul>
      <li>SWCNTs can be added to the model by selecting 'Insert
&gt; Nanotube' and specifying the chirality and length </li>
    </ul>
  </li>
  <li> Graphics:
    <ul>
      <li> In general, rendering speed and quality are improved </li>
      <li> 'View Animation', which animates the model between the
current view and the destination view. This helps maintain model/view
orientation when switching to a standard view (i.e. Front, Top, Home,
etc.) </li>
      <li> 'Help &gt; Graphic Card Info...' displays details
about the system's graphics card </li>
    </ul>
  </li>
  <li> New Jigs:
    <ul>
      <li> 'Measure Distance Jig' displays the distance between
the nuclei and the VdW radii of two atoms </li>
      <li> 'Measure Angle' displays the angle made by three atoms
      </li>
      <li> 'Measure Dihedral' displays the dihedral angle of a
four atom sequence </li>
      <li> 'ESP Image' allows the user to visualize the
electrostatic potential of points on the face of a square 2D surface.
Nano-Hive's MPQC ESP Plane plug-in is used to calculate the
electrostatic potential </li>
      <li> 'Gridplane' allows the user to insert a 2D square or
SiC grid into the part to help measure distances or fuse chunks of SiC </li>
      <li> 'Atom Set' allows the user to group and easily select
arbitrary sets of atoms in the part </li>
    </ul>
  </li>
  <li> Nano-Hive:
    <ul>
      <li> ESP Image plug-in (tested on Windows only) </li>
    </ul>
  </li>
  <li> Fuse Chunks Mode:
    <ul>
      <li> Allows user to fuse two or more chunks together using
one of two methods:
        <ol>
          <li> Make Bonds - bondpoints between chunks will form
bonds when they are near each other </li>
          <li> Fuse Atoms - atoms between chunks will be fused
when they overlap each other </li>
        </ol>
      </li>
    </ul>
  </li>
  <li> Movie Player:
    <ul>
      <li> New 'Loop' option on dashboard restarts the movie when
it reaches the end </li>
      <li> 'POV-Ray Series' supports the 'Skip' value in the
dashboard so that the skipped frames are not saved as POV-Ray files </li>
    </ul>
  </li>
  <li> New Preferences (Edit &gt; Preferences):
    <ul>
      <li> Display Compass Labels </li>
      <li> View Animation
        <ul>
          <li> Animation speed can be changed using a slider
control in the 'Edit &gt; Preferences | General' dialog </li>
          <li> View Animation can be enabled/disabled </li>
        </ul>
      </li>
      <li> Minimization: Watch in real-time </li>
      <li> Bondpoint Highlighting Color </li>
      <li> Ball and Stick Atom Scale </li>
      <li> Ball and Stick Bond Scale </li>
      <li> CPK Atom Scale </li>
      <li> Level of Detail </li>
      <li> Bond Line Thickness </li>
      <li> Startup Mode </li>
      <li> Default Mode </li>
      <li> Display Mode (for each Mode) </li>
      <li> Build Mode Defaults settings, including Autobond,
Water, Highlighting, and Select Atoms of Deposited Object </li>
      <li> Additional Lighting parameters, including Color,
Specular value and light position </li>
      <li> Material Specular Properties </li>
      <li> Nano-Hive Plug-in </li>
      <li> Undo settings, including:
        <ul>
          <li> Restore View when Undoing Structural Changes </li>
          <li> Automatic Checkpoints (default setting) </li>
        </ul>
      </li>
      <li> Window Position and Size Save button </li>
    </ul>
  </li>
  <li> Miscellaneous:
    <ul>
      <li> 'Select &gt; Expand Selection' selects any atom
that is a neighbor of a currently selected atom </li>
      <li> 'Select &gt; Contract Selection' deselects any
atom that is a neighbor of a non-picked atom or has a bondpoint </li>
      <li> 'Modify &gt; Merge' operation extended for atom
selection (i.e. Build Mode) </li>
      <li> 'File &gt; Recent Files' facilitates opening the
most recently used files. </li>
    </ul>
  </li>
</ol>
<hr>
<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in NanoEngineer-1 Version 0.0.6 (8/17/2005):</b>
</p>
<ol>
  <li> Atomic hybridization and higher order bond types:
    <ul>
      <li> Support for the following atomic hybrid orbitals:
        <ul>
          <li> Carbon: sp3, sp2, sp </li>
          <li> Nitrogen: sp3, sp2, sp, graphitic </li>
          <li> Oxygen: sp3, sp2 </li>
          <li> Sulfur: sp3, sp2 </li>
          <li> Hydrogen: s </li>
        </ul>
      </li>
      <li> Support for the following bond types:
        <ul>
          <li> single </li>
          <li> double </li>
          <li> triple </li>
          <li> aromatic </li>
          <li> graphitic </li>
          <li> carbomeric* </li>
        </ul>
* carbomeric bonds are an order-2.5 bond, sometimes used in carbomers.
Alpha 6 supports drawing and valence checking of carbomeric bonds, but
they are simulated and stored in the mmp file as aromatic bonds. This
means carbomeric
bonds can't be stored and reloaded from a mmp file, which is a known
bug we'll
fix for Alpha 7. </li>
      <li> Build Mode:
        <ul>
          <li> New Molecular Modeling Kit (MMKit) used for
selecting elements and atomic hybridizations </li>
          <li> Bonds are selectable and can be deleted using
CTRL+ Left-click (or CMD + Left-click on MacOS X) </li>
          <li> Bonds can be changed using new sub-modes on the
dashboard </li>
          <li> Clipboard integration with the MMKit for copying
and pasting molecules </li>
          <li> Keyboard accelerators for element and atomic
hybridization selections </li>
        </ul>
      </li>
      <li> Select Atoms Mode:
        <ul>
          <li> Support for copying/pasting selected atoms </li>
          <li> Improved dashboard for selecting (and transmuting)
atoms </li>
          <li> Selection Filter now works for Select Connected
and Select Doubly </li>
          <li> Select Connected and Select Doubly have been added
to the dashboard </li>
          <li> Select Doubly much faster </li>
        </ul>
      </li>
      <li> New GAMESS client interface:
        <ul>
          <li> Energy calculations </li>
          <li> Optimizations </li>
          <li> Supports PC GAMESS on Windows </li>
          <li> Supports GAMESS-US on Linux/MacOSX </li>
        </ul>
      </li>
      <li> Jigs:
        <ul>
          <li> Selection of Jigs in the 3D workspace (graphic
area) </li>
          <li> Copying and pasting of jigs (together with their
chunk/atoms) </li>
          <li> New model tree context menu item to select all
atoms of any Jig or set of Jigs </li>
          <li> New model tree context menu item to disable/enable
Jigs, which determines whether the Jig is used during a simulation </li>
          <li> New model tree context menu item for Rotary/Linear
Motors "Align to Chunk", which relaigns the motor to the first chunk it
is attached to </li>
          <li> New model tree context menu item for Rotary/Linear
Motors "Recenter on Atoms", which recenters the motor to the atoms it
is attached to </li>
        </ul>
      </li>
      <li> Molecular Dynamics Simulator:
        <ul>
          <li> Revised Simulator Setup dialog </li>
          <li> Support for Linear motor </li>
          <li> Jigs can be disabled (and enabled) in the model
tree via their context menu </li>
          <li> MD supported for atomic hybridizations and higher
order bonds (see list above) with the following exceptions:
            <ul>
              <li> Still no torsion; this will make systems with
double or aromatic/graphitic bonds act less stiff than they really
should. The parameters for aromatic and graphitic bonds are
interpolated rather than measured directly </li>
              <li> The simulator does not handle hydrogen bonds
or dative bonds </li>
              <li> The parameters for stretches with fluorine and
bends with boron or a double bond to nitrogen are estimated rather than
computed </li>
              <li> The whole set may need an overall correction
factor of some kind. The main variable, absolute energy numbers from
Gamess, is self canceling since we're only using relative energies, and
we used the same level of theory for all the runs. Still there may be
some systematic correction that is needed </li>
              <li> A more sophisticated fitting procedure, and
support for electrostatics, is in the works for Alpha 7 </li>
            </ul>
          </li>
        </ul>
      </li>
      <li> New Preferences dialog (Edit &gt; Preferences):
        <ul>
          <li> atom highlight color </li>
          <li> free valence color </li>
          <li> free valence hotspot color </li>
          <li> bond highlight color </li>
          <li> bond vanes/ribbons color </li>
          <li> bond stretch color </li>
          <li> CPK cylinder color </li>
          <li> higher order bond display modes: multiple
cylinders, vanes, or ribbons </li>
          <li> show valence erros on/off </li>
          <li> turn on/off bond letters for double, triple,
aromatic and graphitic bonds </li>
          <li> background color (solid or gradient) </li>
          <li> compass location </li>
          <li> compass on/off </li>
          <li> origin axis on/off </li>
          <li> point of view axis on/off </li>
        </ul>
      </li>
      <li> Move Chunks:
        <ul>
          <li> New dashboard dropbox includes the following
options:
            <ul>
              <li> Translate </li>
              <li> Rotate X (by specified angle "Theta") </li>
              <li> Rotate Y (by specified angle "Theta") </li>
              <li> Rotate Z (by specified angle "Theta") </li>
            </ul>
          </li>
        </ul>
      </li>
      <li> Movie Player:
        <ul>
          <li> Significant speed improvement </li>
          <li> Circular arrow on end of rotary motor shaft
rotates with motor </li>
          <li> New "POV-Ray Series" save option that allows the
user to dump the current movie as a series of .pov files (which can
then be povrayed by the user and collected into an mpeg offline) </li>
        </ul>
      </li>
      <li> Model Tree:
        <ul>
          <li> New model tree context menu item "Delete all
clipboard items" </li>
          <li> New model tree context menu item "Create new empty
clipboard item" </li>
        </ul>
      </li>
      <li> Opposite View:
        <ul>
          <li> New command (View &gt; Opposite View) that
flips the current view to the opposite side </li>
        </ul>
      </li>
      <li> Align to Common Axis:
        <ul>
          <li> Now aligns selected chunks to the first selected
chunk's axis (instead of the average axis of all selected chunks) </li>
        </ul>
      </li>
    </ul>
  </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in NanoEngineer-1 Version 0.0.5 (5/2/2005):</b>
</p>
<ol>
  <li> Minimize Selection: Will minimize the current selection
only. </li>
  <li> Cookie Cutter:
    <ul>
      <li> Support for Lonsdaleite (hexagonal diamond) </li>
      <li> Modify selection curve(s) for any cookie layer </li>
      <li> Diamond surface orientation displayed in the status
bar </li>
      <li> Rotate lattice grid by increments, around the POV
(point of view) axis. </li>
      <li> Abort/restart the current selection using the Esc key </li>
      <li> Cookie layers are displayed in different colors </li>
      <li> Option to display selection as spheres, similar to CPK
mode, but with bonds drawn as white lines. </li>
      <li> Option to change the grid line color </li>
      <li> Option to display the rest of the model </li>
      <li> New shape selection options (triangle, rectangle,
hexagon, square and diamond) </li>
    </ul>
  </li>
  <li> Clipboard:
    <ul>
      <li> Selected items are now displayed and editable in
Graphics Area. </li>
      <li> Drag and drop works reliably between the Model Tree
and the Clipboard </li>
    </ul>
  </li>
  <li> Move Chunks:
    <ul>
      <li> Translate selected chunks by a specified X, Y, Z
offset </li>
      <li> Move selected chunk(s) to an absolute position </li>
      <li> Dashboard options offer constrained rotation and
translation along the X, Y, or Z axis. </li>
      <li> Modifier keys activate constrained rotation and
translation along the X (X key), Y (Y key) or Z (Z key) axis. </li>
      <li> Modifier keys limit interactive movement to rotation
(R key) or translation (T key) </li>
    </ul>
  </li>
  <li> Fuse Chunks:
    <ul>
      <li> New tool for interactively connecting open bonds
between two or more chunks </li>
      <li> Option to merge connected chunks into a single chunk </li>
      <li> Fine control (using modifier keys) of chunk movement
while finding and highlighting bondable pairs of atom between chunks. </li>
    </ul>
  </li>
  <li> Modify &gt; Invert: Inverts the atom positions of the
selected chunk(s) around their center in space </li>
  <li> Select Atoms: New Atom Selection Filter available on
dashboard </li>
  <li> Simulator:
    <ul>
      <li> Improvements in minimizer, returning the best position
from run </li>
      <li> Multiple runs may improve position further </li>
      <li> Open bonds are treated as if they were hydrogenated. </li>
      <li> More informative history messages for the user after a
simulation or minimization completes. </li>
    </ul>
  </li>
  <li> New Select Menu: Includes all options from the Select
Toolbar </li>
  <li> Display &gt; Show Invisible Atoms: Resets the display
of any invisible atoms in the selected chunk(s). </li>
  <li> Display &gt; Reset Chunk Display: Resets the display
of all atoms in the selected chunk(s). </li>
  <li> Plot Tool: Changed GNUplot filename suffix from
"-trace.plt" to "-plot.txt" </li>
  <li> Removed Element Selector from Modify toolbar. </li>
  <li> Added Element Selector to the Build Atoms dashboard. </li>
  <li> Lighting Tool: Added Restore Defaults button. </li>
  <li> Changed Modify &gt; Weld to Modify &gt; Merge </li>
  <li> Hide/Unhide state of Model Tree nodes are saved in the MMP
file and restored when reopening the part. </li>
  <li> Open/close state of Model Tree groups are saved in MMP
file and restored when reopening the part. </li>
  <li> Jigs:
    <ul>
      <li> Jigs can be disabled using the Model Tree context
menu, or by dragging them to the Clipboard; disabled jigs remain
attached to the same atoms, even when saved to and reloaded from an MMP
file, but have no effect on the simulator until they are re-enabled. </li>
      <li> Jigs can be located anywhere in the Model Tree, and
this location is not altered when saving the MMP file. </li>
    </ul>
  </li>
</ol>
<hr>
<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in v0.0.4 (3/14/2005):</b>
</p>
<ol>
  <li> Element Selector:
    <ul>
      <li> removed controls for changing element colors. </li>
      <li> Refined buttons to be easier to read, especially on
MacOSX. </li>
    </ul>
  </li>
  <li> Element Color Settings:
    <ul>
      <li> new control panel for changing element colors. </li>
      <li> accessed by "Display &gt; Element Color
Settings..." </li>
      <li> element colors saved to nE-1 preferences database </li>
      <li> element colors can be saved to a text file in human
readable format </li>
      <li> load element colors from a file </li>
      <li> can restore default nE-1 element colors </li>
    </ul>
  </li>
  <li> Lighting: New Lighting control panel for changing lighting
settings. </li>
  <li> Plot Tool:
    <ul>
      <li> new control panel for plotting simulator trace file
output using GNUplot. </li>
      <li> buttons on control panel open trace and plot files in
a text editor. </li>
      <li> plot files can be edited by hand and replotted using
"Plot" button </li>
    </ul>
  </li>
  <li> Simulator trace files now created next to DPB trajectory
file, with the name dpb_filename-trace.txt </li>
  <li> Simulator: new toolbar/menu contains Simulator, Movie
Player and Plot Tool. </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in v0.0.3 (2/28/2005):</b>
</p>
<ol>
  <li> Select &gt; Doubly: The cursor displays the "Wait
Cursor" (hourglass) while Select &gt; Doubly is working. </li>
  <li> Model Tree: Improved Drag and Drop behavour in the Model
Tree widget. </li>
  <li> Clipboard: Fixed some bugs related to the Clipboard. More
stable now. </li>
  <li> Movie Player: Program no longer displays errors when
selecting Done while movie is playing. </li>
  <li> Element Selector: Users can change element colors for all
supported element types. </li>
  <li> Element Selector: Users can load/save custom element
colors from/to a file. </li>
  <li> Extrude Tool: The bond-offset spheres are turned off by
default. They can be turned on from the dashboard. </li>
  <li> Extrude Tool: The bond-offset spheres are rendered in a
manner to clue the user on which configurations will produce the most
bonds. </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in v0.0.2 (2/14/2005):</b>
</p>
<ol>
  <li> Simulator works on MacOSX. </li>
  <li> Fixed bug in cad that limited the simulator to no more
than 940 frames on Windows. </li>
</ol>
<hr>
<!-- +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
<p><b>What was new in v0.0.1 (2/2/2005):</b>
</p>
<ol>
  <li> Everything </li>
</ol>
</body>
</html>
